### PR TITLE
WIP Typescript lualib

### LINKED ISSUE #80: typescript_lualib.lua is generated in root directory when running tests
We have to figure out which test causes that and remove/fix that test.

Its ignored so it won't appear in the repo, but its still annoying.

My guess is that translation tests create the file.

### LINKED ISSUE #134: Runtime Library should only be imported for files which require it.
If a file contains no references the Runtime library (Set, Map, Array.push, etc.), the library should not be required.

### PR DESCRIPTION
**Pull request by lolleko**
Open PR to reimplement runtime lualib features in typescript.

Fixes #80, Closes #134 

**TODO** Compiler option for different import modes `--luaLibImport`
- [x] `require` require a bundled lualib in each file basically the same we currently have
- [x] `inline` inline the used lualib functions add the bottom of the file. (Probably default option)
- [x] `noImport` don't import or inline any runtime lib features

this will replaces the existing `--dontRequireLuaLib`

**TODO** Reimplement all lualib features in TS

- [x] ArrayEvery
- [x] ArrayFilter
- [x] ArrayForEach
- [x] ArrayIndexOf
- [x] ArrayMap
- [x] ArrayPush
- [x] ArraySlice
- [x] ArraySome
- [x] ArraySplice
- [x] InstanceOf
- [x] Map
- [x] Set
- [x] StringReplace
- [x] StringSplit
- [x] Ternary

### CODE REVIEW

File: src/CommandLineParser.ts [hunk 1]
```diff
@@ -18,16 +18,11 @@ export class CLIError extends Error {
...
-        type: "boolean",
-    },
-    dontRequireLuaLib: {
-        default: false,
-        describe: "Dont require lua library that enables advanced Typescipt/JS functionality.",
-        type: "boolean",
+    luaLibImport: {
+        choices: ["inline", "require", "none"],
+        default: "inline",
+        describe: "Specify Lua target version.",
```
Comment (Perryvw) [inline src/CommandLineParser.ts]: Incorrect description.

File: src/Compiler.ts [hunk 2]
```diff
@@ -49,6 +51,8 @@ export function compile(fileNames: string[], options: CompilerOptions): void {
             try {
                 const rootDir = options.rootDir;
 
+                // console.log(`Transpiling ${sourceFile.fileName}...`);
```
Comment (Perryvw) [inline src/Compiler.ts]: Either uncomment or remove (I like it to see which file is causing errors).

File: src/Index.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env node
+
+import {compile} from "./Compiler";
```
Comment (Perryvw) [inline src/Index.ts]: Is this file linted? Would expect `{ compile }`

File: src/Transpiler.ts [hunk 4]
```diff
@@ -106,11 +135,17 @@ export abstract class LuaTranspiler {
 
     public makeExports(): string {
         let result = "";
-        this.exportStack.pop().forEach(exp => result += this.makeExport(exp.name, exp.node, exp.dummy));
+        this.exportStack.pop().forEach(
+            exp => result += this.makeExport(exp.name, exp.node, exp.dummy));
```
Comment (Perryvw) [inline src/Transpiler.ts]: random newline?

File: src/Transpiler.ts [hunk 5]
```diff
@@ -106,11 +135,17 @@ export abstract class LuaTranspiler {
...
         return result;
     }
 
-    public getAbsouluteImportPath(relativePath: string) {
+    public importLuaLibFeature(feature: LuaLibFeature) {
```
Comment (Perryvw) [inline src/Transpiler.ts]: Add return type (is there a ts-lint rule for this?)

File: src/Transpiler.ts [hunk 6]
```diff
@@ -106,11 +135,17 @@ export abstract class LuaTranspiler {
...
+        // TODO inline imported features in output i option set
```
Comment (Perryvw) [inline src/Transpiler.ts]: This is still a TODO?

File: src/Transpiler.ts [hunk 7]
```diff
@@ -106,11 +135,17 @@ export abstract class LuaTranspiler {
...
+        this.luaLibFeatureSet.add(feature);
+    }
+
+    public getAbsoluteImportPath(relativePath: string) {
```
Comment (Perryvw) [inline src/Transpiler.ts]: Return type.

File: src/Transpiler.ts [hunk 8]
```diff
@@ -251,19 +302,20 @@ export abstract class LuaTranspiler {
             let result = `local ${fileImportTable} = require(${resolvedImportPath})\n`;
             this.importCount++;
             imports.elements.forEach(element => {
-                const nameText = element.name.escapedText;
+                const nameText = this.transpileIdentifier(element.name);
                 if (element.propertyName) {
+                    const propertyText = this.transpileIdentifier(element.propertyName);
                     result +=
```
Comment (Perryvw) [inline src/Transpiler.ts]: Newline still required here? Kinda ugly imo

File: src/Transpiler.ts [hunk 9]
```diff
@@ -251,19 +302,20 @@ export abstract class LuaTranspiler {
...
-                        `local ${nameText} = ${fileImportTable}.${element.propertyName.escapedText}\n`;
+                        `local ${nameText} = ${fileImportTable}.${propertyText}\n`;
                 } else {
                     result +=
```
Comment (Perryvw) [inline src/Transpiler.ts]: Get rid of newline if possible.

File: src/Transpiler.ts [hunk 10]
```diff
@@ -1164,6 +1240,29 @@ export abstract class LuaTranspiler {
         }
     }
 
+    public transpileIdentifier(identifier: ts.Identifier): string {
```
Comment (Perryvw) [inline src/Transpiler.ts]: We should probably add a comment with explanation + link to relevant TS docs/code

File: src/Transpiler.ts [hunk 11]
```diff
@@ -1164,6 +1240,29 @@ export abstract class LuaTranspiler {
...
+            && escapedText.charCodeAt(1) === underScoreCharCode
+            && escapedText.charCodeAt(2) === underScoreCharCode) {
+            return escapedText.substr(1);
+        }
+        return escapedText;
+    }
+
+    public transpileArrayBindingElement(name: ts.ArrayBindingElement): string {
+        if (ts.isOmittedExpression(name)) {
+            return "_";
```
Comment (Perryvw) [inline src/Transpiler.ts]: Should probably be `__` since some people use `_` for lambda variables.

File: src/Transpiler.ts [hunk 12]
```diff
@@ -1175,7 +1274,9 @@ export abstract class LuaTranspiler {
 
         node.declarationList.declarations.forEach(declaration => {
             result += this.transpileVariableDeclaration(declaration as ts.VariableDeclaration);
-            this.pushExport((declaration.name as ts.Identifier).escapedText, node);
+            if (ts.isIdentifier((declaration.name as ts.Identifier))) {
```
Comment (Perryvw) [inline src/Transpiler.ts]: Too many brackets.

File: src/Transpiler.ts [hunk 13]
```diff
@@ -1201,8 +1302,8 @@ export abstract class LuaTranspiler {
             }
 
             const vars = node.name.elements.map(
-                    element => ((element as ts.BindingElement).name as ts.Identifier
-                ).escapedText).join(",");
+                    element =>
+                        (this.transpileIdentifier((element as ts.BindingElement).name as ts.Identifier))).join(",");
```
Comment (Perryvw) [inline src/Transpiler.ts]: transpileArrayBindingElement seems to fit here.

File: src/lualib/ArraySplice.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        for (let k = len - actualDeleteCount; k > actualStart; k--) {
+            const from = k + actualDeleteCount - 1;
+            const to = k + itemCount - 1;
+
+            if (list[from]) {
+                list[to] = list[from];
+            } else {
+                list[to] = undefined;
+            }
+
```
Comment (Perryvw) [inline src/lualib/ArraySplice.ts]: Extraneous whiteline

File: src/lualib/ArraySplice.ts [hunk 15]
```diff
@@ -0,0 +1,93 @@
+// https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-262.pdf
+function __TS__ArraySplice<T>(list: T[], start: number, deleteCount: number, ...items: T[]): T[] {
+    // 1. 2.
```
Comment (Perryvw) [inline src/lualib/ArraySplice.ts]: I would either write out the pseudo-code in the comments or just remove them. These numbers don't really add anything imo.

File: src/lualib/Set.ts [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+            }
+        } else if (other !== undefined) {
+            this.size = other.length;
+            for (const value of other) {
+                this.items[value as any] = true as any;
+            }
+        }
+    }
+
+    public add(value: TValue) {
```
Comment (Perryvw) [inline src/lualib/Set.ts]: Return type

File: test/unit/lualib.spec.ts [hunk 17]
```diff
@@ -204,20 +204,26 @@ export class LuaLibArrayTests {
     @TestCase([], "test1")
```
Comment (Perryvw) [inline test/unit/lualib.spec.ts]: This file should probably also be moved to test/unit/lualib

File: src/Compiler.ts (hunk 2 above)
Comment (lolleko) [reply src/Compiler.ts]: It should probably only print on error then.

**Pull request merged**